compilers/concrete-optimizer/concrete-optimizer-cpp/src/concrete-optimizer.rs

```diff
@@ -852,15 +662,54 @@ impl Dag {
             ffi::MultiParamStrategy::ByPrecision | _ => PartitionCut::for_each_precision(&self.0),
         };
         let circuit_sol =
-            concrete_optimizer::optimization::dag::multi_parameters::optimize_generic::optimize(
-                &self.0,
-                config,
-                &search_space,
-                encoding,
-                options.default_log_norm2_woppbs,
-                &caches_from(options),
-                &Some(p_cut),
-            );
+            if !options.keyset_restriction.is_null() && !options.range_restriction.is_null() {
```

Comment:
Unfortunately not an equivalent alternative. The functiono is generic over the restriction type, but this means that for every different restriction type I must have a different call. I could use dynamic dyspatch with `dyn SearchSpaceRestriction`, but this would mean having dynamic dispatch for every call to `restriction.is_available` ... Since it is called quite a lot, I think we are better off with the big switch case for now. We may reconsider if we get more complex restrictions.